src/DataFactory.js

```diff
@@ -1,182 +1,47 @@
-const FateInt = require('./types/FateInt')
-const FateBool = require('./types/FateBool')
-const FateString = require('./types/FateString')
-const FateHash = require('./types/FateHash')
-const FateSignature = require('./types/FateSignature')
-const FateList = require('./types/FateList')
-const FateMap = require('./types/FateMap')
-const FateTuple = require('./types/FateTuple')
-const FateVariant = require('./types/FateVariant')
-const FateBytes = require('./types/FateBytes')
-const FateBits = require('./types/FateBits')
-const FateAccountAddress = require('./types/FateAccountAddress')
-const FateContractAddress = require('./types/FateContractAddress')
-const FateChannelAddress = require('./types/FateChannelAddress')
-const FateOracleAddress = require('./types/FateOracleAddress')
-const FateOracleQueryAddress = require('./types/FateOracleQueryAddress')
+const assert = require('./utils/assert')
+const PrimitiveDataFactory = require('./DataFactory/PrimitiveDataFactory')
+const ListDataFactory = require('./DataFactory/ListDataFactory')
+const MapDataFactory = require('./DataFactory/MapDataFactory')
+const TupleDataFactory = require('./DataFactory/TupleDataFactory')
+const RecordDataFactory = require('./DataFactory/RecordDataFactory')
+const VariantDataFactory = require('./DataFactory/VariantDataFactory')
 
 const zip = (arr, ...arrs) => {
     return arr.map((val, i) => arrs.reduce((a, curr) => [...a, curr[i]], [val]))
 }
 
 class DataFactory {
-    constructor(aci) {
-        this.aci = aci
+    #factories
+
+    #mapper
+
+    constructor(mapper) {
+        this.#mapper = mapper
+        this.#factories = [
+            new PrimitiveDataFactory(this),
+            new ListDataFactory(this),
+            new MapDataFactory(this),
+            new TupleDataFactory(this),
+            new RecordDataFactory(this),
+            new VariantDataFactory(this),
+        ]
     }
 
     create(types, values) {
-        const requiredTypesCount = types
-            .map(type => FateVariant.isFateTypeOption(type))
-            .lastIndexOf(false) + 1
-        if (values.length > types.length || values.length < requiredTypesCount) {
-            throw new Error(
-                'Non matching number of arguments. '
-                + `Got ${values.length} but expected from ${requiredTypesCount} to ${types.length}`
-            )
-        }
+        assert(
+            values.length === types.length,
```

Comment:
This check duplicates the one in Encoder.js: https://github.com/aeternity/aepp-calldata-js/pull/110/files#diff-2bd04d4b33af87492499134517e1ce3fe0cdd9ccafd7337075df34bf1114a172R62-R66

Reply:
It is intentional.